### PR TITLE
[CRT] atexit.c: Fix memory handling

### PR DESCRIPTION
**Pull request by SergeGautherie**
And related improvements.

Addendum to b30bf6de315557f80cfa16574689563dfe08eab5 (r54651).

### CODE REVIEW

**Comment by learn-more** [comment]
What does it 'fix' other than breaking the previously zero-initialized table?

**Comment by SergeGautherie** [comment]
> What does it 'fix' other than breaking the previously zero-initialized table?

How does it break?
Especially, what good does zeroing do? As the first part is overwritten and the second part is not accessed...

File: sdk/lib/crt/stdlib/atexit.c [hunk 1]
```diff
@@ -69,16 +79,14 @@ _onexit_t CDECL _onexit(_onexit_t func)
   {
     _onexit_t *newtable;
     TRACE("expanding table\n");
-    newtable = calloc(atexit_table_size + 32, sizeof(void *));
+    newtable = realloc(atexit_table, (atexit_table_size + 32) * sizeof(_onexit_t));
```
Comment (ThFabba) [inline sdk/lib/crt/stdlib/atexit.c:82]: Missing zero-init

Comment (SergeGautherie) [reply sdk/lib/crt/stdlib/atexit.c:82]: Again, what is zero-init needed for?

Comment (ThFabba) [reply sdk/lib/crt/stdlib/atexit.c:82]: My mistake, didn't see there's `atexit_registered` in addition to `atexit_table_size`.

**Comment by ThFabba** [comment]
Your branch name mentions cppcheck -- so is there an actual cppcheck warning you're fixing here? If so please quote/explain it. If not, this seems entirely unnecessary.

**Comment by SergeGautherie** [comment]
> Wine sync

Wine code is different, though I could try to upstream bits that would apply there...
https://source.winehq.org/git/wine.git/blob/17529582402ebe27ef975fc7dcb8353f4f95e629:/dlls/msvcrt/exit.c#l65

> actual cppcheck warning

* https://qarmin.gitlab.io/-/reactos/-/jobs/577525577/artifacts/report/1011.html#line-72
* '[CRT] _onexit(): Use the explicit '_onexit_t' type' commit.

**Comment by ThFabba** [comment]
> Wine code is different, though I could try to upstream bits that would apply there...
> https://source.winehq.org/git/wine.git/blob/17529582402ebe27ef975fc7dcb8353f4f95e629:/dlls/msvcrt/exit.c#l65

Wine code has simply changed over time.
https://source.winehq.org/git/wine.git/commitdiff/83f8b226948d40a81917a649b7ef83e2ea46a845


>     * https://qarmin.gitlab.io/-/reactos/-/jobs/577525577/artifacts/report/1011.html#line-72
>     * '[CRT] _onexit(): Use the explicit '_onexit_t' type' commit.

So cppcheck is smart enough to know that `newtable` is pointer-sized but not that `*newtable` is pointer-sized? That sounds like a flat-out bug in their analyzer.
In any case, it may be worth silencing the warning... for which that single commit looks just fine. Doesn't warrant a complete refactor of upstream code.

**Comment by SergeGautherie** [comment]
> code has simply changed

Thanks for the link. I'll look at that separately.

> just fix the bug

Removed 2 refactor commits.

**Pull request merged**